File: native/app/contribution_content.go
package app

import (
	"errors"
	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/sites"
)

func contributionManifestPolicy(m core.Manifest) bool {
	if m.Kind != "site-contribution" || m.PublicKey != nil || len(m.Keys) < 1 || len(m.Keys) > 2 || m.Expires-m.Created > sites.ContributionLifetimeMS {
		return false
	}
	for _, key := range m.Keys {
		if key.Reader == m.Author.ID {
			return true
		}
	}
	return false
}
func inspectContributionBundle(bundle core.Bundle, identity *core.Identity) (map[string]any, error) {
	if bundle.Manifest.Kind != "site-contribution" {
		return nil, nil
	}
	if !contributionManifestPolicy(bundle.Manifest) {
		return nil, errors.New("propostas exigem um envelope privado limitado")
	}
	readable := false
	for _, key := range bundle.Manifest.Keys {
		if identity != nil && key.Reader == identity.Public.ID {
			readable = true
		}
	}
	if !readable {
		return nil, nil
	}
	value, err := core.DecryptBundle(bundle, identity)
	if err != nil {
		return nil, err
	}
	return sites.MatchContributionEnvelope(bundle, value)
}

File: apps/node/src/contribution-content.ts
import {
  decryptBundle,
  type Bundle,
  type Identity,
} from "../../../packages/core/src/index";
import { nodeCertificateCrypto } from "../../../packages/core/src/certificate-crypto";
import { contributionManifestPolicy } from "../../../packages/content/src/site-contribution";
import { createContributionEnvelopeProtocol } from "../../../packages/sites/src/contribution-envelope";
const envelopes = createContributionEnvelopeProtocol(nodeCertificateCrypto);
export function inspectContribution(bundle: Bundle, identity?: Identity) {
  if (bundle.manifest.kind !== "site-contribution") return null;
  if (!contributionManifestPolicy(bundle.manifest))
    throw Error("Propostas exigem um envelope privado limitado");
  if (
    !identity ||
    !bundle.manifest.keys.some((k) => k.reader === identity.public.id)
  )
    return null;
  return envelopes.match(bundle, decryptBundle(bundle, identity));
}
